### PR TITLE
Promise based transaction draining logic

### PR DESCRIPTION
**Pull request by seglo**
The draining control for Transactional sources will occasionally get into a situation where the consumer source can no longer be signalled by the producer that a transaction is aborted, in order to flush in flight messages used to drain messages on shutdown.

Should fix https://github.com/akka/alpakka-kafka/issues/1108

Other improvements:

* Add first observed offset to transactional batch for troubleshooting with transaction commit/abort logging
* Prettyprint transactional batch in logging for transaction commit and abort
* Cancel commit timer when `onCompletionFailure` is fired (not sure if this is necessary since we don't `setKeepGoing(true)`, but it's also done on `onCompletionSuccess` and it can't hurt.
* ~~Add `onFirstMessageReceived` callback.  If a transactional source is shutdown while messages are inflight, but the first message hasn't been produced, then we don't need to drain inflight messages.~~
  * This doesn't work.  In cases messages don't reach the producer before the consumer shuts down (due to failure or a stream completion signal) then in-flight messages will eventually be produced and committed by `TransactionalProducerStage` anyway.
* Add maximum restart condition to integration tests to avoid transient live lock scenarios where we continuously abort transactions and can't finish test.

### CODE REVIEW

**Comment by seglo** [comment]
I haven't been able to reproduce this flakiness on master.

**Comment by seglo** [comment]
Nevermind, I just did, though it seems to happen less frequently.  The symptom is that the transactional source gets stuck in a loop checking for inflight messages to be committed or aborted, but neither ever occurs because the messages from the producer are never received.

**Comment by ennru** [comment]
This doesn't really make it simpler, does it?
But you seem to be on to something here...

**Comment by seglo** [comment]
Looks like I broke exactly once consistency checks somehow.  I'll take a look at some point today, but this might not be ready for 2.0.3.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 1]
```diff
@@ -38,31 +40,40 @@ private[kafka] final class TransactionalProducerStage[K, V, P](
...
     // There is no guarantee that offsets adding callbacks will be called in any particular order.
     // Decreasing an offset stored for the KTP would mean possible data duplication.
     // Since `awaitingConfirmation` counter guarantees that all writes finished, we can safely assume
     // that all all data up to maximal offsets has been wrote to Kafka.
     private val previousHighest = tail.getOrElse(head.key, -1L)
     private[internal] val offsets = tail + (head.key -> head.offset.max(previousHighest))
+    private[internal] lazy val firstOffsetMap = first.get(head.key) match {
+      case None => first + (head.key -> head.offset)
+      case _ => first
+    }
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:76]: I could probably merge this into a single map of first and end offsets.

File: core/src/main/scala/akka/kafka/internal/TransactionalSources.scala [hunk 2]
```diff
@@ -379,6 +414,18 @@ private final class TransactionalSubSourceStageLogic[K, V](
...
+    onTransactionAborted.future.onComplete {
+      case Success(_) => onTransactionAbortedCb.invoke(())
+      case TFailure(ex) => log.error(ex, "An error occurred while during the transaction abort process")
+    }(materializer.executionContext)
+    onFirstMessageReceived.future.onComplete {
+      case Success(_) => onFirstMessageReceivedCb.invoke(())
+      case TFailure(ex) => log.error(ex, "The first message was not received")
+    }(materializer.executionContext)
+  }
+
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/TransactionalSources.scala:428]: Would be nice to consolidate this with the partitioned sources.

File: core/src/main/scala/akka/kafka/internal/TransactionalSources.scala [hunk 3]
```diff
@@ -413,12 +460,8 @@ private final class TransactionalSubSourceStageLogic[K, V](
     case (sender, Committed(offsets)) =>
       inFlightRecords.committed(offsets.view.mapValues(_.offset() - 1).toMap)
       sender ! Done
-    case (sender, CommittingFailure) => {
-      log.info("Committing failed, resetting in flight offsets")
-      inFlightRecords.reset()
-    }
     case (sender, Drain(partitions, ack, msg)) =>
-      if (inFlightRecords.empty(partitions)) {
+      if (!firstMessageReceived || inFlightRecords.empty(partitions)) {
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/TransactionalSources.scala:464]: If a message is never received downstream then we don't need to drain.

File: tests/src/it/scala/akka/kafka/TransactionsPartitionedSourceSpec.scala [hunk 4]
```diff
@@ -33,6 +33,12 @@ class TransactionsPartitionedSourceSpec extends SpecBase
   with Repeated {
 
   val replicationFactor = 2
+  // It's possible to get into a livelock situation where the `restartAfter` interval causes transactions to abort
+  // over and over.  This can happen when there are a few partitions left to process and they can never be fully
+  // processed because we always restart the stream before the transaction can be completed successfully.
+  // The `maxRestarts` provides an upper bound for the maximum number of times we restart the stream so if we get
+  // into a livelock it can eventually be resolved by not restarting any more.
+  val maxRestarts = new AtomicInteger(1000)
```
Comment (seglo) [inline tests/src/it/scala/akka/kafka/TransactionsPartitionedSourceSpec.scala:41]: This situation was occurring before and after this fix.  This should make the tests _always_ complete, it just places an upper bound on the number of restarts.

File: tests/src/it/scala/akka/kafka/TransactionsSourceSpec.scala [hunk 5]
```diff
@@ -55,13 +63,16 @@ class TransactionsSourceSpec extends SpecBase
...
+        (0 until sourcePartitions).map { part =>
+          val rangeStart = ((part * partitionSize) + 1)
+          val rangeEnd = (partitionSize * (part + 1))
+          log.info(s"Producing [$rangeStart to $rangeEnd] to partition $part")
+          produce(sourceTopic, rangeStart to rangeEnd, part)
+        }
 
       Await.result(Future.sequence(producers), 1.minute)
 
-      val consumerSettings = consumerDefaults.withGroupId(group)
```
Comment (seglo) [inline tests/src/it/scala/akka/kafka/TransactionsSourceSpec.scala:64]: This change should be rolled back.

File: tests/src/test/scala/akka/kafka/TransactionsOps.scala [hunk 6]
```diff
@@ -162,11 +182,10 @@ trait TransactionsOps extends TestSuite with Matchers {
       .plainSource(settings, Subscriptions.topics(topic))
       .map(r => (r.partition(), r.offset(), r.value()))
       .take(elementsToTake)
-      .idleTimeout(30.seconds)
       .alsoTo(
         Flow[(Int, Long, String)]
           .scan(0) { case (count, _) => count + 1 }
-          .filter(_ % 100 == 0)
+          .filter(_ % 10 == 0)
```
Comment (seglo) [inline tests/src/test/scala/akka/kafka/TransactionsOps.scala:188]: rollback

File: tests/src/it/scala/akka/kafka/TransactionsSourceSpec.scala [hunk 7]
```diff
@@ -101,7 +119,6 @@ class TransactionsSourceSpec extends SpecBase
 
       val consumer = offsetValueSource(probeConsumerSettings(probeConsumerGroup), sinkTopic)
         .take(elements.toLong)
-        .idleTimeout(30.seconds)
```
Comment (seglo) [inline tests/src/it/scala/akka/kafka/TransactionsSourceSpec.scala:104]: This is a magic number that can sometimes cause the test to abort too early before the transactional stream had a chance to complete.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 8]
```diff
@@ -37,47 +40,61 @@ private[kafka] final class TransactionalProducerStage[K, V, P](
...
-    // Since `awaitingConfirmation` counter guarantees that all writes finished, we can safely assume
-    // that all all data up to maximal offsets has been wrote to Kafka.
-    private val previousHighest = tail.getOrElse(head.key, -1L)
-    private[internal] val offsets = tail + (head.key -> head.offset.max(previousHighest))
+
+    private val withHighest = tail.get(head.key) match {
+      // There is no guarantee that offsets adding callbacks will be called in any particular order.
+      // Decreasing an offset stored for the KTP would mean possible data duplication.
+      // Since `awaitingConfirmation` counter guarantees that all writes finished, we can safely assume
+      // that all all data up to maximal offsets has been written to Kafka.
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:74]:
```suggestion
      // that all data up to maximal offsets has been written to Kafka.
```

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 9]
```diff
@@ -88,8 +105,16 @@ private object TransactionalProducerStage {
...
     }
-  }
 
+    override def toString: String = {
+      val header = tpFirstEndOffsetFormat.format("TopicPartition", "First Observed Offset", "End Offset")
+      val rows = offsets.map {
+        case (gtp, FirstAndHighestOffsets(first, highest)) =>
+          tpFirstEndOffsetFormat.format(gtp.topicPartition.toString, first, highest)
+      }
+      "\n" + header + rows.mkString
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:115]: Does this need to be multiline?

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 10]
```diff
@@ -228,25 +259,36 @@ private final class TransactionalProducerStageLogic[K, V, P](
...
-    batchOffsets.committingFailed()
-    super.onCompletionFailure(ex)
+
+    batchOffsets
+      .transactionAborted()
+      .onComplete {
+        case Success(_) =>
+          log.debug("Successfully sent source stage failure")
+          onCommitFailedAckCb.invoke(ex)
+        case Failure(commitEx) if NonFatal(commitEx) =>
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:271]: No need to check for `NonFatal`, futures wouldn't pass fatal exceptions.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 11]
```diff
@@ -228,25 +259,36 @@ private final class TransactionalProducerStageLogic[K, V, P](
...
+      }(materializer.executionContext)
   }
 
   private def commitTransaction(batch: NonemptyTransactionBatch, beginNewTransaction: Boolean): Unit = {
     val group = batch.group
     log.debug("Committing transaction for transactional id '{}' consumer group '{}' with offsets: {}",
               transactionalId,
               group,
-              batch.offsets)
+              batch.toString)
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:282]:
```suggestion
              batch)
```

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 12]
```diff
@@ -228,25 +259,36 @@ private final class TransactionalProducerStageLogic[K, V, P](
...
     val offsetMap = batch.offsetMap()
     producer.sendOffsetsToTransaction(offsetMap.asJava, group)
     producer.commitTransaction()
     log.debug("Committed transaction for transactional id '{}' consumer group '{}' with offsets: {}",
               transactionalId,
               group,
-              batch.offsets)
+              batch.toString)
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:289]:
```suggestion
              batch)
```

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 13]
```diff
@@ -258,6 +300,12 @@ private final class TransactionalProducerStageLogic[K, V, P](
     }
   }
 
+  private val onCommitFailedAckCb: AsyncCallback[Throwable] = {
+    getAsyncCallback[Throwable](
+      t => super.onCompletionFailure(t)
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:305]: The `super` is hideous, should it be just `failStage(t)`?

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala [hunk 14]
```diff
@@ -275,7 +323,7 @@ private final class TransactionalProducerStageLogic[K, V, P](
   }
 
   private def abortTransaction(reason: String): Unit = {
-    log.debug("Aborting transaction: {}", reason)
+    log.debug("Aborting transaction with reason: {}, offsets: {}", reason, batchOffsets.toString)
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:326]:
```suggestion
    log.debug("Aborting transaction with reason: {}, offsets: {}", reason, batchOffsets)
```

File: core/src/main/scala/akka/kafka/internal/TransactionalSources.scala [hunk 15]
```diff
@@ -139,22 +144,36 @@ private[internal] abstract class TransactionalSourceLogic[K, V, Msg](shape: Sour
 
   override val groupId: String = consumerSettings.properties(ConsumerConfig.GROUP_ID_CONFIG)
 
+  private val onTransactionAbortedCb = getAsyncCallback[Unit] { _ =>
+    log.debug(s"Committing failed, draining: $draining, resetting inFlightRecords: $inFlightRecords")
```
Comment (ennru) [inline core/src/main/scala/akka/kafka/internal/TransactionalSources.scala:148]:
```suggestion
    log.debug("Committing failed, draining: {}, resetting inFlightRecords: {}", draining, inFlightRecords")
```

Is this the only use of `draining`?

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala (hunk 9 above)
Comment (seglo) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:115]: Yes, it produces a table.

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala (hunk 10 above)
Comment (seglo) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:271]: :+1:

File: core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala (hunk 13 above)
Comment (seglo) [reply core/src/main/scala/akka/kafka/internal/TransactionalProducerStage.scala:305]: We could, but if we ever implement anything in `DefaultProducerStage.onCompleteFailure` we may want it to execute that.

File: core/src/main/scala/akka/kafka/internal/TransactionalSources.scala (hunk 15 above)
Comment (seglo) [reply core/src/main/scala/akka/kafka/internal/TransactionalSources.scala:148]: Yes.  It was for troubleshooting.  I don't think it's very useful anymore.  I'll remove it.

**Comment by seglo** [comment]
New failure: https://travis-ci.org/github/akka/alpakka-kafka/jobs/691735521#L386

**Comment by hepin1989** [comment]
@seglo Friendly ping, Any update?

**Comment by ennru** [comment]
This turned out harder than it should be, unless someone in the community digs into it a lot, we won't be able to finish this PR.

**Comment by patriknw** [comment]
I think we have to close this attempt

**Pull request closed without merging**